### PR TITLE
[SYCL] Remove `generic_type_lists.hpp`

### PR DESCRIPTION
**Pull request by aelovikov-intel**
We don't need to instantiate all the `vec`/`marray` types to implement `generic_type_traits.hpp`.

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/experimental/cuda/builtins.hpp [hunk 1]
```diff
@@ -25,37 +25,15 @@ namespace oneapi {
...
-                            sycl::detail::gtl::scalar_floating_list,
-                            sycl::detail::gtl::scalar_signed_integer_list,
-                            sycl::detail::gtl::scalar_unsigned_integer_list>;
-} // namespace detail
-
 template <typename T>
 inline __SYCL_ALWAYS_INLINE std::enable_if_t<
-    sycl::detail::is_contained<
-        T, sycl::ext::oneapi::experimental::cuda::detail::ldg_types>::value,
+    detail::check_type_in_v<detail::element_type_t<T>, char, signed char, short,
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/experimental/cuda/builtins.hpp:30]: I think this part isn't NFC in a sense that `gtl::scalar_floating_list` on the left included `bfloat16` but the code in this hasn't been adjusted to actually handle it when it was added there. As such, the change here is actually good.

Comment (maarquitos14) [reply sycl/include/sycl/ext/oneapi/experimental/cuda/builtins.hpp:30]: Should we leave a TODO reminding it on code?

**Review by maarquitos14 — APPROVED** [review]
LGTM, just one nit.

File: sycl/include/sycl/ext/oneapi/experimental/cuda/builtins.hpp (hunk 1 above)
Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/experimental/cuda/builtins.hpp:33]: IMO, that's up to the @intel/llvm-reviewers-cuda to decide if any changes are necessary here.

File: sycl/include/sycl/detail/generic_type_traits.hpp [hunk 2]
```diff
@@ -28,51 +27,66 @@ namespace sycl {
...
 template <typename T>
-inline constexpr bool is_genfloat_v = is_contained_v<T, gtl::floating_list>;
+inline constexpr bool is_svgenfloath_v =
+    std::is_same_v<T, half> ||
+    (is_vec_v<T> && std::is_same_v<element_type_t<T>, half>);
 
 template <typename T>
 inline constexpr bool is_sgenfloat_v =
-    is_contained_v<T, gtl::scalar_floating_list>;
+    check_type_in_v<T, float, double, half, ext::oneapi::bfloat16>;
```
Comment (rolandschulz) [inline sycl/include/sycl/detail/generic_type_traits.hpp:51]: given that you replace `is_contained_v` with `check_type_in_v` its implementation should probably be changed to the `is_base_of` way of checking membership for best performance

Comment (aelovikov-intel) [reply sycl/include/sycl/detail/generic_type_traits.hpp:51]: Yeah, I'll look into that separately.

**Comment by aelovikov-intel** [comment]
@npmiller , @intel/llvm-reviewers-cuda , ping, you're the last approval needed.

**Review by npmiller — APPROVED** [review]
CUDA changes LGTM

**Pull request merged**